File: Labs/Lab-04-20231013/app/src/main/java/sap/escooters/data_source_layer/DataSourceLayerImpl.java
package sap.escooters.data_source_layer;

import java.io.*;
import java.util.Optional;
import io.vertx.core.json.JsonObject;
import sap.layers.Layer;

public class DataSourceLayerImpl implements DataSourceLayer {

	private String USERS_PATH = "users";
	private String ESCOOTERS_PATH = "escooters";
	private String RIDES_PATH = "rides";
	
	private String dbaseFolder;
	
	public DataSourceLayerImpl(String dbaseFolder) {
		this.dbaseFolder =  dbaseFolder;
	}

	@Override
	public void init(Optional<Layer> layer) {
		makeDir(dbaseFolder);
		makeDir(dbaseFolder + File.separator + USERS_PATH);
		makeDir(dbaseFolder + File.separator + ESCOOTERS_PATH);
		makeDir(dbaseFolder + File.separator + RIDES_PATH);
	}	

	@Override
	public void saveUser(JsonObject user) throws DataSourceException {
		this.saveObj(USERS_PATH, user.getString("id"), user);
	}

	@Override
	public void saveEScooter(JsonObject escooter) throws DataSourceException {
		this.saveObj(ESCOOTERS_PATH, escooter.getString("id"), escooter);
	}

	@Override
	public void saveRide(JsonObject ride) throws DataSourceException {
		this.saveObj(RIDES_PATH, ride.getString("id"), ride);
	}

	private void saveObj(String db, String id, JsonObject obj) throws DataSourceException {
		try {									
			FileWriter fw = new FileWriter(dbaseFolder + File.separator + db + File.separator + id + ".json");
			java.io.BufferedWriter wr = new BufferedWriter(fw);				
			wr.write(obj.encodePrettily());
			wr.flush();
			fw.close();
		} catch (Exception ex) {
			ex.printStackTrace();
			throw new DataSourceException(ex.getMessage());
		}
	}
	
	private void makeDir(String name) {
		try {
			File dir = new File(name);
			if (!dir.exists()) {
				dir.mkdir();
			}
		} catch (Exception ex) {
			ex.printStackTrace();
		}
	}

	
}

File: Labs/Lab-05-20231016/app/src/main/kotlin/assigment03/escooters/data_source_layer/DataSourceLayerImpl.kt
package assigment03.escooters.data_source_layer

import assigment03.layers.Layer
import io.vertx.core.json.JsonObject
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter

class DataSourceLayerImpl(private val dbaseFolder: String) : DataSourceLayer {
    private val userPath = "users"
    private val escooterPath = "escooters"
    private val riderPath = "rides"

    override fun init(layer: Layer?) {
        makeDir(dbaseFolder)
        makeDir("$dbaseFolder${File.separator}$userPath")
        makeDir("$dbaseFolder${File.separator}$escooterPath")
        makeDir("$dbaseFolder${File.separator}$riderPath")
    }

    @Throws(DataSourceException::class)
    override fun saveUser(user: JsonObject) = saveObj(userPath, user.getString("id"), user)

    @Throws(DataSourceException::class)
    override fun saveEScooter(escooter: JsonObject) = saveObj(escooterPath, escooter.getString("id"), escooter)

    @Throws(DataSourceException::class)
    override fun saveRide(ride: JsonObject) = saveObj(riderPath, ride.getString("id"), ride)

    private fun saveObj(db: String, id: String, obj: JsonObject) {
        try {
            FileWriter("$dbaseFolder${File.separator}$db${File.separator}$id.json").use { fw ->
                BufferedWriter(fw).use { wr ->
                    wr.write(obj.encodePrettily())
                }
            }
        } catch (ex: Exception) {
            ex.printStackTrace()
            throw DataSourceException(ex.message ?: "Unknown Error")
        }
    }

    private fun makeDir(name: String) {
        try {
            File(name).takeIf { !it.exists() }?.mkdir()
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }
}